### PR TITLE
OPCT-5 - plugin: add cluster-upgrade plugin

### PR DESCRIPTION
**Pull request by mtulio**
Add cluster-upgrade plugin to trigger the upgrade using `openshift-tests` utility with option `run-upgrade`.

https://issues.redhat.com/browse/SPLAT-651

Blocked by PRs:

- [x] https://github.com/redhat-openshift-ecosystem/provider-certification-plugins/pull/29
- [x] https://github.com/redhat-openshift-ecosystem/provider-certification-plugins/pull/30
- [x] https://github.com/redhat-openshift-ecosystem/provider-certification-plugins/pull/31

Blocking CLI:

- https://github.com/redhat-openshift-ecosystem/provider-certification-tool/pull/33

### CODE REVIEW

File: openshift-tests-provider-cert/plugin/global_env.sh [hunk 1]
```diff
@@ -35,6 +35,26 @@ declare -grx UTIL_OTESTS_BIN="${SHARED_DIR}/openshift-tests"
 declare -grx UTIL_OTESTS_READY="${SHARED_DIR}/openshift-tests.ready"
 declare -grx UTIL_OTESTS_FAILED="${SHARED_DIR}/openshift-tests.failed"
 
+# PLUGIN Instances
+declare -grx PLUGIN_ID_KUBE_CONFORMANCE="10"
+declare -grx PLUGIN_NAME_KUBE_CONFORMANCE="${PLUGIN_ID_OPENSHIFT_CONFORMANCE}-openshift-kube-conformance"
```
Comment (bostrt) [inline openshift-tests-provider-cert/plugin/global_env.sh:40]: Change `PLUGIN_ID_OPENSHIFT_CONFORMANCE` to `PLUGIN_ID_KUBE_CONFORMANCE`

File: openshift-tests-provider-cert/plugin/executor.sh [hunk 2]
```diff
@@ -25,6 +45,11 @@ os_log_info "[executor] Executor started. Choosing execution type based on envir
 if [[ -n "${CERT_TEST_SUITE}" ]]; then
     os_log_info "Starting openshift-tests suite [${CERT_TEST_SUITE}] Provider Conformance executor..."
 
+    os_log_info_local "WORKAROUND: chaning SCC..."
+    oc adm policy add-scc-to-group privileged system:authenticated system:serviceaccounts || true
+    oc adm policy add-scc-to-group anyuid system:authenticated system:serviceaccounts || true
+
```
Comment (mtulio) [inline openshift-tests-provider-cert/plugin/executor.sh:51]: should be removed, not sure or guarantee that will work as expected

File: openshift-tests-provider-cert/plugin/global_fn.sh [hunk 3]
```diff
@@ -41,18 +41,22 @@ init_config() {
...
     elif [[ "${PLUGIN_ID:-}" == "${PLUGIN_ID_OPENSHIFT_CONFORMANCE}" ]]
     then
-        PLUGIN_NAME="20-openshift-conformance-validated"
-        CERT_TEST_SUITE="openshift/conformance"
-        PLUGIN_BLOCKED_BY+=("10-openshift-kube-conformance")
+        PLUGIN_NAME="${PLUGIN_NAME_OPENSHIFT_UPGRADE}"
+        PLUGIN_BLOCKED_BY=()
 
     else
         os_log_info "[init_config] Unknow value for PLUGIN_ID=[${PLUGIN_ID:-}]"
```
Comment (bostrt) [inline openshift-tests-provider-cert/plugin/global_fn.sh:62]: Typo `Unknow value` -> `Unknown value`

File: openshift-tests-provider-cert/plugin/global_fn.sh [hunk 4]
```diff
@@ -41,18 +41,22 @@ init_config() {
...
+    elif [[ "${PLUGIN_ID:-}" == "${PLUGIN_ID_OPENSHIFT_CONFORMANCE}" ]]
     then
-        PLUGIN_NAME="10-openshift-kube-conformance"
-        CERT_TEST_SUITE="kubernetes/conformance"
-        PLUGIN_BLOCKED_BY=()
+        PLUGIN_NAME="${PLUGIN_NAME_OPENSHIFT_CONFORMANCE}"
+        CERT_TEST_SUITE="${OPENSHIFT_TESTS_SUITE_OPENSHIFT_CONFORMANCE}"
+        PLUGIN_BLOCKED_BY+=("${PLUGIN_NAME_KUBE_CONFORMANCE}")
 
     elif [[ "${PLUGIN_ID:-}" == "${PLUGIN_ID_OPENSHIFT_CONFORMANCE}" ]]
```
Comment (bostrt) [inline openshift-tests-provider-cert/plugin/global_fn.sh:56]: Change `PLUGIN_ID_OPENSHIFT_CONFORMANCE` to `PLUGIN_ID_OPENSHIFT_UPGRADE`

File: openshift-tests-provider-cert/plugin/global_fn.sh [hunk 5]
```diff
@@ -41,18 +41,22 @@ init_config() {
...
+        CERT_TEST_SUITE="${OPENSHIFT_TESTS_SUITE_OPENSHIFT_CONFORMANCE}"
+        PLUGIN_BLOCKED_BY+=("${PLUGIN_NAME_KUBE_CONFORMANCE}")
+
+    elif [[ "${PLUGIN_ID:-}" == "${PLUGIN_ID_OPENSHIFT_UPGRADE}" ]]
+    then
+        PLUGIN_NAME="${PLUGIN_NAME_OPENSHIFT_UPGRADE}"
+        PLUGIN_BLOCKED_BY=()
 
     else
         os_log_info "[init_config] Unknow value for PLUGIN_ID=[${PLUGIN_ID:-}]"
```
Comment (mtulio) [inline openshift-tests-provider-cert/plugin/global_fn.sh:62]:
```suggestion
        os_log_info "[init_config] Unknown value for PLUGIN_ID=[${PLUGIN_ID:-}]"
```

File: openshift-tests-provider-cert/plugin/global_env.sh (hunk 1 above)
Comment (mtulio) [reply openshift-tests-provider-cert/plugin/global_env.sh]: fixed after rebase

**Comment by mtulio** [comment]
Converting to draft as I am still working on it.

File: openshift-tests-provider-cert/plugin/executor.sh [hunk 6]
```diff
@@ -7,16 +7,35 @@
 #TODO(mtulio): pipefail should be commented until we provide a better solution
 # to handle errors (failed e2e) on sub-process managed by openshift-tests main proc.
 # https://issues.redhat.com/browse/SPLAT-592
-#set -o pipefail
-set -o nounset
+# set -o pipefail
 # set -o errexit
+set -o nounset
+
```
Comment (mtulio) [inline openshift-tests-provider-cert/plugin/executor.sh:13]: un{releated,necessary} change

Comment (mtulio) [reply openshift-tests-provider-cert/plugin/executor.sh:13]: https://issues.redhat.com/browse/SPLAT-599

File: openshift-tests-provider-cert/plugin/global_env.sh [hunk 7]
```diff
@@ -40,10 +40,37 @@ declare -grx UTIL_OC_BIN="/usr/bin/oc"
...
+declare -grx PLUGIN_RUN_MODE_UPGRADE="upgrade"
+
+# Sonobuot Plugin Statuses
+declare -grx SONOBUOY_PLUGIN_STATUS_COMPLETE="complete"
+declare -grx SONOBUOY_PLUGIN_STATUS_FAILED="failed"
+
+## Setting 3h for the plugin blocker feature (wait-plugin).
+## We don't want to run forever, however starting prematurely is not acceptable.
+declare -grx PLUGIN_WAIT_TIMEOUT_COUNT=1080
+declare -grx PLUGIN_WAIT_TIMEOUT_INTERVAL=10
```
Comment (mtulio) [inline openshift-tests-provider-cert/plugin/global_env.sh:73]: https://issues.redhat.com/browse/OPCT-15

File: openshift-tests-provider-cert/plugin/global_fn.sh [hunk 8]
```diff
@@ -1,11 +1,11 @@
...
-os_log_info() {
-    echo "#$(caller |awk '{print$2":"$1}')> " "$@"
-}
-export -f os_log_info
+# # os_log_info logger function, printing the current bash script
+# # and line as prefix.
+# os_log_info() {
+#     echo "#$(caller |awk '{print$2":"$1}')> " "$@"
+# }
+# export -f os_log_info
```
Comment (mtulio) [inline openshift-tests-provider-cert/plugin/global_fn.sh:8]: Changes must be added isolated

File: openshift-tests-provider-cert/plugin/global_fn.sh [hunk 9]
```diff
@@ -93,10 +97,17 @@ show_config() {
 PLUGIN_NAME=${PLUGIN_NAME}
 PLUGIN_BLOCKED_BY=${PLUGIN_BLOCKED_BY[*]}
 PLUGIN_ID=${PLUGIN_ID}
+CERT_LEVEL=${CERT_LEVEL:-''}
```
Comment (mtulio) [inline openshift-tests-provider-cert/plugin/global_fn.sh:100]: https://issues.redhat.com/browse/OPCT-16

File: openshift-tests-provider-cert/plugin/report-progress.sh [hunk 10]
```diff
@@ -127,19 +159,19 @@ watch_dependency_done() {
...
                 timeout_checks=0
-                sleep 10
+                sleep "${PLUGIN_WAIT_TIMEOUT_INTERVAL}"
                 continue
             fi
             # dont run timeouts for if blocker is waiting
             if [[ "${blocker_msg}" =~ "status=waiting-for" ]] && [[ "${state_blocking}" =~ "blocked-by" ]]; then
                 timeout_checks=0
-                sleep 10
+                sleep "${PLUGIN_WAIT_TIMEOUT_INTERVAL}"
```
Comment (mtulio) [inline openshift-tests-provider-cert/plugin/report-progress.sh:174]: https://issues.redhat.com/browse/OPCT-15

File: openshift-tests-provider-cert/plugin/report-progress.sh [hunk 11]
```diff
@@ -180,15 +214,27 @@ report_progress() {
...
+                    # Stop flooding the aggregator API with long error messages, when
+                    # there are many errors on the cluster. Note: this will not affect
+                    # the final results, it's just the progress API.
+                    if [[ ${err_count} -gt 30 ]]; then
+                        PROGRESS["failures"]+=,\"$(echo "$line" | cut -d"\"" -f2)\"
+                    else
+                        sig_err=$(echo "$line" | cut -d"\"" -f2 | grep -Po '^(\[sig-[a-zA-Z-]*\])')
+                        err_msg="${sig_err} ERR#${err_count}"
+                        PROGRESS["failures"]+=,\"${err_msg}\"
+                    fi
```
Comment (mtulio) [inline openshift-tests-provider-cert/plugin/report-progress.sh:232]: https://issues.redhat.com/browse/OPCT-24

File: openshift-tests-provider-cert/plugin/wait-plugin.sh [hunk 12]
```diff
@@ -81,7 +81,7 @@ if [[ ${#PLUGIN_BLOCKED_BY[@]} -ge 1 ]]; then
                 exit 1
             fi
             os_log_info_local "Waiting 30s for Plugin[${bl_plugin_name}]...[${timeout_checks}/${timeout_count}]"
-            sleep 30
+            sleep "${PLUGIN_WAIT_TIMEOUT_INTERVAL}"
```
Comment (mtulio) [inline openshift-tests-provider-cert/plugin/wait-plugin.sh:84]: https://issues.redhat.com/browse/OPCT-15

**Review by mtulio — COMMENTED** [review]
Quickly overview of the ready PR to support upgrade feature.

**Comment by mtulio** [comment]
Hey robert and richard, ptal?
/assign @bostrt @rvanderp3

**Comment by mtulio** [comment]
Tests with MCP check - forcing to fail the plugin when it is not present on RUN_MODE=upgrade:

1) Without MCP:
- Plugins (05,10,20) fail fast with custom failed JUnit
- Artifacts plugin (99) runs (data is collected), regardless the upgrade and conformance plugins failing


<details>
  <summary>Step-by-step (click to expand)</summary>
  
  ```text
$ KUBECONFIG=$PWD/.opct-4120/clusters/opct-4120/auth/kubeconfig $CLI_BIN_PATH run -w --mode upgrade --upgrade-to-image=$(oc adm release info 4.13.0-ec.2 -o jsonpath={.image}) --dev-count 10
INFO[2023-01-24T17:10:10-03:00] Ensuring proper node label for dedicated mode exists 
INFO[2023-01-24T17:10:11-03:00] Ensuring the tool will run in the privileged environment... 
INFO[2023-01-24T17:10:11-03:00] Created opct-scc-privileged ClusterRole      
INFO[2023-01-24T17:10:11-03:00] Created opct-scc-privileged ClusterRoleBinding 
INFO[2023-01-24T17:10:11-03:00] Running OpenShift Provider Certification Tool... 
INFO[2023-01-24T17:10:12-03:00] object already exists                         name=openshift-provider-certification namespace= resource=namespaces
INFO[2023-01-24T17:10:12-03:00] create request issued                         name=sonobuoy-config-cm namespace=openshift-provider-certification resource=configmaps
INFO[2023-01-24T17:10:12-03:00] create request issued                         name=sonobuoy-plugins-cm namespace=openshift-provider-certification resource=configmaps
INFO[2023-01-24T17:10:12-03:00] create request issued                         name=sonobuoy namespace=openshift-provider-certification resource=pods
INFO[2023-01-24T17:10:13-03:00] create request issued                         name=sonobuoy-aggregator namespace=openshift-provider-certification resource=services
INFO[2023-01-24T17:10:13-03:00] Jobs scheduled! Waiting for resources be created... 
Tue, 24 Jan 2023 17:10:25 -03> Global Status: running
JOB_NAME                           | STATUS     | RESULTS    | PROGRESS                  | MESSAGE                                           
05-openshift-cluster-upgrade       | running    |            |                           |                                                   
10-openshift-kube-conformance      | running    |            |                           |                                                   
20-openshift-conformance-validated | running    |            |                           |                                                   
99-openshift-artifacts-collector   | running    |            |                           |                                                   
Tue, 24 Jan 2023 17:10:36 -03> Global Status: running
JOB_NAME                           | STATUS     | RESULTS    | PROGRESS                  | MESSAGE                                           
05-openshift-cluster-upgrade       | complete   |            |                           | waiting for post-processor...                     
10-openshift-kube-conformance      | complete   |            |                           | waiting for post-processor...                     
20-openshift-conformance-validated | complete   |            |                           | waiting for post-processor...                     
99-openshift-artifacts-collector   | running    |            |                           |                                                   
(...)

INFO[2023-01-24T17:22:36-03:00] Waiting for post-processor...                
Tue, 24 Jan 2023 17:25:56 -03> Global Status: complete
JOB_NAME                           | STATUS     | RESULTS    | PROGRESS                  | MESSAGE                                           
05-openshift-cluster-upgrade       | complete   | failed     |                           | Total tests processed: 1 (0 pass / 1 failed)      
10-openshift-kube-conformance      | complete   | failed     |                           | Total tests processed: 1 (0 pass / 1 failed)      
20-openshift-conformance-validated | complete   | failed     |                           | Total tests processed: 1 (0 pass / 1 failed)      
99-openshift-artifacts-collector   | complete   | passed     | 0/0 (0 failures)          | Total tests processed: 3 (3 pass / 0 failed)      
INFO[2023-01-24T17:25:56-03:00] The execution has completed! Use retrieve command to collect the results and share the archive with your Red Hat partner. 

$ KUBECONFIG=$PWD/.opct-4120/clusters/opct-4120/auth/kubeconfig $CLI_BIN_PATH results 202301242010_sonobuoy_ac9b2894-55ad-475e-94e1-c715ee60486b.tar.gz
Plugin: 10-openshift-kube-conformance
Status: failed
Total: 1
Passed: 0
Failed: 1
Skipped: 0

Failed tests:
[opct][moode=upgrade] MachineConfigPool opct not found. Return code=1

Plugin: 20-openshift-conformance-validated
Status: failed
Total: 1
Passed: 0
Failed: 1
Skipped: 0

Failed tests:
[opct][moode=upgrade] MachineConfigPool opct not found. Return code=1

Plugin: 99-openshift-artifacts-collector
Status: passed
Total: 3
Passed: 3
Failed: 0
Skipped: 0

Plugin: 05-openshift-cluster-upgrade
Status: failed
Total: 1
Passed: 0
Failed: 1
Skipped: 0

Failed tests:
[opct][moode=upgrade] MachineConfigPool opct not found. Return code=1

  ```
</details>


2) With MCP (4.12.0->4.13.0-ec.2)
- Cluster is upgraded successfull
- Conformance tests will run normally (output below is using DEV_MODE - limited number of e2e to validate the flow)

<details>
  <summary>Step-by-step (click to expand)</summary>

  ```text

$ KUBECONFIG=$PWD/.opct-4120/clusters/opct-4120/auth/kubeconfig oc get mcp
NAME     CONFIG                                             UPDATED   UPDATING   DEGRADED   MACHINECOUNT   READYMACHINECOUNT   UPDATEDMACHINECOUNT   DEGRADEDMACHINECOUNT   AGE
master   rendered-master-db05047a395e586bd650a0f9b54ecc3f   True      False      False      3              3                   3                     0                      6h50m
opct                                                                                                                                                                        13m
worker   rendered-worker-ee636661f68fea596813db895f764179   True      False      False      3              3                   3                     0                      6h50m

$ KUBECONFIG=$PWD/.opct-4120/clusters/opct-4120/auth/kubeconfig $CLI_BIN_PATH run -w --mode upgrade --upgrade-to-image=$(oc adm release info 4.13.0-ec.2 -o jsonpath={.image}) --dev-count 10
INFO[2023-01-24T19:05:29-03:00] Ensuring proper node label for dedicated mode exists 
INFO[2023-01-24T19:05:29-03:00] Ensuring the tool will run in the privileged environment... 
INFO[2023-01-24T19:05:29-03:00] Created opct-scc-privileged ClusterRole      
INFO[2023-01-24T19:05:29-03:00] Created opct-scc-privileged ClusterRoleBinding 
INFO[2023-01-24T19:05:29-03:00] Running OpenShift Provider Certification Tool... 
INFO[2023-01-24T19:05:30-03:00] object already exists                         name=openshift-provider-certification namespace= resource=namespaces
INFO[2023-01-24T19:05:30-03:00] create request issued                         name=sonobuoy-config-cm namespace=openshift-provider-certification resource=configmaps
INFO[2023-01-24T19:05:31-03:00] create request issued                         name=sonobuoy-plugins-cm namespace=openshift-provider-certification resource=configmaps
INFO[2023-01-24T19:05:31-03:00] create request issued                         name=sonobuoy namespace=openshift-provider-certification resource=pods
INFO[2023-01-24T19:05:31-03:00] create request issued                         name=sonobuoy-aggregator namespace=openshift-provider-certification resource=services
INFO[2023-01-24T19:05:31-03:00] Jobs scheduled! Waiting for resources be created... 
(...)                                                 
Tue, 24 Jan 2023 19:06:14 -03> Global Status: running
JOB_NAME                           | STATUS     | RESULTS    | PROGRESS                  | MESSAGE                                           
05-openshift-cluster-upgrade       | running    |            | 0/0 (0 failures)          | status=4.12.0=upgrade-progressing-False           
10-openshift-kube-conformance      | running    |            | 0/359 (0 failures)        | status=waiting-for=05-openshift-cluster-upgrade=(0/0/0)=[0/1080]
20-openshift-conformance-validated | running    |            | 0/3454 (0 failures)       | status=waiting-for=10-openshift-kube-conformance=(0/0/0)=[0/1080]
99-openshift-artifacts-collector   | running    |            | 0/0 (0 failures)          | status=waiting-for=20-openshift-conformance-validated=(0/0/0)=[0/1080]
(...)
Tue, 24 Jan 2023 19:06:44 -03> Global Status: running
JOB_NAME                           | STATUS     | RESULTS    | PROGRESS                  | MESSAGE                                           
05-openshift-cluster-upgrade       | running    |            | 0/0 (0 failures)          | status=Working towards 4.13.0-ec.2: 12 of 831 done (1% complete)
10-openshift-kube-conformance      | running    |            | 0/359 (0 failures)        | status=waiting-for=05-openshift-cluster-upgrade=(0/0/0)=[3/1080]
20-openshift-conformance-validated | running    |            | 0/3454 (0 failures)       | status=blocked-by=10-openshift-kube-conformance=(0/-359/0)=[0/1080]
99-openshift-artifacts-collector   | running    |            | 0/0 (0 failures)          | status=blocked-by=20-openshift-conformance-validated=(0/-3454/0)=[0/1080]
(...)
Tue, 24 Jan 2023 20:52:34 -03> Global Status: post-processing
JOB_NAME                           | STATUS     | RESULTS    | PROGRESS                  | MESSAGE                                           
05-openshift-cluster-upgrade       | complete   |            | 0/0 (0 failures)          | waiting for post-processor...                     
10-openshift-kube-conformance      | complete   |            | 0/359 (0 failures)        | waiting for post-processor...                     
20-openshift-conformance-validated | complete   |            | 0/3454 (0 failures)       | waiting for post-processor...                     
99-openshift-artifacts-collector   | complete   |            | 0/0 (0 failures)          | waiting for post-processor...                     
INFO[2023-01-24T20:52:34-03:00] Waiting for post-processor...                
Tue, 24 Jan 2023 20:55:54 -03> Global Status: complete
JOB_NAME                           | STATUS     | RESULTS    | PROGRESS                  | MESSAGE                                           
05-openshift-cluster-upgrade       | complete   | failed     | 0/0 (0 failures)          | Total tests processed: 38 (37 pass / 1 failed)    
10-openshift-kube-conformance      | complete   | failed     | 0/359 (0 failures)        | Total tests processed: 13 (12 pass / 1 failed)    
20-openshift-conformance-validated | complete   | failed     | 0/3454 (0 failures)       | Total tests processed: 7 (6 pass / 1 failed)      
99-openshift-artifacts-collector   | complete   | passed     | 0/0 (0 failures)          | Total tests processed: 3 (3 pass / 0 failed)      
INFO[2023-01-24T20:55:54-03:00] The execution has completed! Use retrieve command to collect the results and share the archive with your Red Hat partner. 

$ KUBECONFIG=$PWD/.opct-4120/clusters/opct-4120/auth/kubeconfig $CLI_BIN_PATH results 202301242205_sonobuoy_980f8d7b-9b69-442b-9243-9ef497455826.tar.gz
Plugin: 05-openshift-cluster-upgrade
Status: failed
Total: 38
Passed: 37
Failed: 1
Skipped: 0

Failed tests:
[sig-mco] Machine config pools complete upgrade

Plugin: 10-openshift-kube-conformance
Status: failed
Total: 13
Passed: 12
Failed: 1
Skipped: 0

Failed tests:
[sig-arch] Monitor cluster while tests execute

Plugin: 20-openshift-conformance-validated
Status: failed
Total: 13
Passed: 6
Failed: 1
Skipped: 6

Failed tests:
[sig-arch] Monitor cluster while tests execute

Plugin: 99-openshift-artifacts-collector
Status: passed
Total: 3
Passed: 3
Failed: 0
Skipped: 0

$ KUBECONFIG=$PWD/.opct-4120/clusters/opct-4120/auth/kubeconfig oc get clusterversion version -o json |jq .status.history
[
  {
    "acceptedRisks": "Forced through blocking failures: Multiple precondition checks failed:\n* Precondition \"EtcdRecentBackup\" failed because of \"ControllerStarted\": RecentBackup: The etcd backup controller is starting, and will decide if recent backups are available or if a backup is required\n* Precondition \"ClusterVersionRecommendedUpdate\" failed because of \"UnknownUpdate\": RetrievedUpdates=True (), so the update from 4.12.0 to 4.13.0-ec.2 is probably neither recommended nor supported.",
    "completionTime": "2023-01-24T23:13:07Z",
    "image": "quay.io/openshift-release-dev/ocp-release@sha256:01192353b3c3e536779cfa0fc910064299df15ce01be0cff7188868588d32321",
    "startedTime": "2023-01-24T22:06:32Z",
    "state": "Completed",
    "verified": true,
    "version": "4.13.0-ec.2"
  },
  {
    "completionTime": "2023-01-24T15:36:32Z",
    "image": "quay.io/openshift-release-dev/ocp-release@sha256:4c5a7e26d707780be6466ddc9591865beb2e3baa5556432d23e8d57966a2dd18",
    "startedTime": "2023-01-24T15:06:14Z",
    "state": "Completed",
    "verified": false,
    "version": "4.12.0"
  }
]

  ```
</details>

**Comment by mtulio** [comment]
@rvanderp3 this PR is ready for review. Ptal?
/assign @rvanderp3

**Comment by mtulio** [comment]
Tested upgrades from 4.12 to 4.13 successfully:

```bash
$ $CLI_BIN_PATH version
OpenShift Provider Certification Tool: 0.0.0+40f5965
Sonobuoy Version: v0.56.10,

$ jq . .opct-4120/clusters/opct-4120-2023012601/opct/results/resources/ns/openshift-provider-certification/core_v1_pods.json  |grep 'quay.io/ocp-cert/openshift-tests-provider-cert:v' |uniq -c
     16             "image": "quay.io/ocp-cert/openshift-tests-provider-cert:v0.3.0-alpha0",

$ jq -r '.items[].status.history | map({"startTime": .startedTime , "completeTime": .completionTime, "version": .version}) ' .opct-4120/clusters/opct-4120-2023012601/opct/results/resources/cluster/config.openshift.io_v1_clusterversions.json 
[
  {
    "startTime": "2023-01-25T20:53:52Z",
    "completeTime": "2023-01-25T21:58:57Z",
    "version": "4.13.0-ec.2"
  },
  {
    "startTime": "2023-01-25T17:46:00Z",
    "completeTime": "2023-01-25T18:16:30Z",
    "version": "4.12.0"
  }
]



 $ ./openshift-provider-cert results .opct-4120/clusters/opct-4120-2023012601/opct/202301252051_sonobuoy_574e2b07-ed25-4e75-8a48-f910027afb04.tar.gz |grep ^Plugin -A 5
Plugin: 05-openshift-cluster-upgrade
Status: failed
Total: 38
Passed: 37
Failed: 1
Skipped: 0
--
Plugin: 10-openshift-kube-conformance
Status: failed
Total: 704
Passed: 684
Failed: 20
Skipped: 0
--
Plugin: 20-openshift-conformance-validated
Status: failed
Total: 3753
Passed: 1655
Failed: 45
Skipped: 2053
--
Plugin: 99-openshift-artifacts-collector
Status: passed
Total: 3
Passed: 3
Failed: 0
Skipped: 0


```

**Pull request merged**